// react-static/dist/static.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.buildXMLandRSS = exports.writeRoutesToStatic = exports.getConfig = undefined;

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var buildXMLandRSS = exports.buildXMLandRSS = function () {
  var _ref7 = _asyncToGenerator( /*#__PURE__*/_regenerator2.default.mark(function _callee3(_ref8) {
    var config = _ref8.config;
    var xml, generateXML;
    return _regenerator2.default.wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            generateXML = function generateXML(_ref9) {
              var routes = _ref9.routes;

              var xml = '<?xml version="1.0" encoding="UTF-8"?><urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">';
              routes.forEach(function (route) {
                if (route.noindex) {
                  return;
                }
                xml += '<url>';
                xml += '<loc>' + route.permalink + '</loc>';
                xml += route.lastModified ? '<lastmod>' + route.lastModified + '</lastmod>' : '';
                xml += route.changeFreq ? '<changefreq>' + route.changeFreq + '</changefreq>' : '';
                xml += route.priority ? '<priority>' + route.priority + '</priority>' : '';
                xml += '</url>';
              });
              xml += '</urlset>';
              return xml;
            };

            if (config.siteRoot) {
              _context3.next = 4;
              break;
            }

            console.log('\n=> Warning: No \'siteRoot\' defined in \'static.config.js\'!\n=> This is required for both absolute url\'s and a sitemap.xml to be exported.\n');
            return _context3.abrupt('return');

          case 4:
            xml = generateXML({
              routes: config.routes.map(function (route) {
                return _extends({
                  permalink: _path2.default.join(config.siteRoot, route.path),
                  changeFreq: 600000,
                  lastModified: '',
                  priority: 0.5
                }, route);
              })
            });
            _context3.next = 7;
            return _fsExtra2.default.writeFile(_path2.default.join(_paths.DIST, 'sitemap.xml'), xml);

          case 7:
          case 'end':
            return _context3.stop();
        }
      }
    }, _callee3, this);
  }));

  return function buildXMLandRSS(_x3) {
    return _ref7.apply(this, arguments);
  };
}();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _server = require('react-dom/server');

var _fsExtra = require('fs-extra');

var _fsExtra2 = _interopRequireDefault(_fsExtra);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _reactHelmet = require('react-helmet');

var _reactHelmet2 = _interopRequireDefault(_reactHelmet);

var _DefaultHtml = require('./DefaultHtml');

var _DefaultHtml2 = _interopRequireDefault(_DefaultHtml);

var _paths = require('./paths');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; } /* eslint-disable import/no-dynamic-require, react/no-danger, no-cond-assign */

//


var defaultEntry = './src/index';

var getConfig = exports.getConfig = function getConfig() {
  return require(_path2.default.resolve(_path2.default.join(process.cwd(), 'static.config.js'))).default;
};

var writeRoutesToStatic = function () {
  var _ref = _asyncToGenerator( /*#__PURE__*/_regenerator2.default.mark(function _callee2(_ref2) {
    var config = _ref2.config;
    var userConfig, HtmlTemplate, Comp;
    return _regenerator2.default.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            userConfig = getConfig();
            HtmlTemplate = config.Html || _DefaultHtml2.default;
            Comp = require(_path2.default.join(_paths.ROOT, userConfig.entry || defaultEntry)).default;
            return _context2.abrupt('return', Promise.all(config.routes.map(function () {
              var _ref3 = _asyncToGenerator( /*#__PURE__*/_regenerator2.default.mark(function _callee(route) {
                var initialProps, URL, InitialPropsContext, ContextualComp, staticMeta, appHtml, helmet, head, Html, Head, Body, html, htmlFilename, initialPropsFilename, writeHTML, writeJSON;
                return _regenerator2.default.wrap(function _callee$(_context) {
                  while (1) {
                    switch (_context.prev = _context.next) {
                      case 0:
                        _context.t0 = route.getProps;

                        if (!_context.t0) {
                          _context.next = 5;
                          break;
                        }

                        _context.next = 4;
                        return route.getProps({ route: route, dev: false });

                      case 4:
                        _context.t0 = _context.sent;

                      case 5:
                        initialProps = _context.t0;

                        if (!initialProps) {
                          _context.next = 9;
                          break;
                        }

                        _context.next = 9;
                        return _fsExtra2.default.outputFile(_path2.default.join(_paths.DIST, route.path, 'routeData.json'), JSON.stringify(initialProps || {}));

                      case 9:
                        URL = route.path;

                        // Inject initialProps into static build

                        InitialPropsContext = function (_Component) {
                          _inherits(InitialPropsContext, _Component);

                          function InitialPropsContext() {
                            _classCallCheck(this, InitialPropsContext);

                            return _possibleConstructorReturn(this, (InitialPropsContext.__proto__ || Object.getPrototypeOf(InitialPropsContext)).apply(this, arguments));
                          }

                          _createClass(InitialPropsContext, [{
                            key: 'getChildContext',
                            value: function getChildContext() {
                              return {
                                initialProps: initialProps,
                                URL: URL
                              };
                            }
                          }, {
                            key: 'render',
                            value: function render() {
                              return this.props.children;
                            }
                          }]);

                          return InitialPropsContext;
                        }(_react.Component);

                        InitialPropsContext.childContextTypes = {
                          initialProps: _propTypes2.default.object,
                          URL: _propTypes2.default.string
                        };
                        ContextualComp = _react2.default.createElement(
                          InitialPropsContext,
                          null,
                          Comp
                        );
                        staticMeta = {};

                        if (!config.preRenderMeta) {
                          _context.next = 22;
                          break;
                        }

                        _context.t1 = _extends;
                        _context.t2 = {};
                        _context.t3 = staticMeta;
                        _context.next = 20;
                        return config.preRenderMeta(ContextualComp);

                      case 20:
                        _context.t4 = _context.sent;
                        staticMeta = (0, _context.t1)(_context.t2, _context.t3, _context.t4);

                      case 22:
                        appHtml = (0, _server.renderToString)(ContextualComp);

                        // Extract head calls using Helmet

                        helmet = _reactHelmet2.default.renderStatic();
                        head = {
                          htmlProps: helmet.htmlAttributes.toComponent(),
                          bodyProps: helmet.bodyAttributes.toComponent(),
                          base: helmet.base.toComponent(),
                          link: helmet.link.toComponent(),
                          meta: helmet.meta.toComponent(),
                          noscript: helmet.noscript.toComponent(),
                          script: helmet.script.toComponent(),
                          style: helmet.style.toComponent(),
                          title: helmet.title.toComponent()
                        };


                        staticMeta.head = head;

                        if (!config.postRenderMeta) {
                          _context.next = 34;
                          break;
                        }

                        _context.t5 = _extends;
                        _context.t6 = {};
                        _context.t7 = staticMeta;
                        _context.next = 32;
                        return config.postRenderMeta(appHtml);

                      case 32:
                        _context.t8 = _context.sent;
                        staticMeta = (0, _context.t5)(_context.t6, _context.t7, _context.t8);

                      case 34:
                        Html = function Html(_ref4) {
                          var children = _ref4.children,
                              rest = _objectWithoutProperties(_ref4, ['children']);

                          return _react2.default.createElement(
                            'html',
                            _extends({ lang: 'en' }, head.htmlprops, rest),
                            children
                          );
                        };

                        Head = function Head(_ref5) {
                          var children = _ref5.children,
                              rest = _objectWithoutProperties(_ref5, ['children']);

                          return _react2.default.createElement(
                            'head',
                            rest,
                            head.base,
                            head.link,
                            head.meta,
                            head.noscript,
                            head.script,
                            head.style,
                            head.title,
                            children
                          );
                        };

                        Body = function Body(_ref6) {
                          var children = _ref6.children,
                              rest = _objectWithoutProperties(_ref6, ['children']);

                          return _react2.default.createElement(
                            'body',
                            _extends({}, head.bodyProps, rest),
                            children,
                            _react2.default.createElement('script', {
                              type: 'text/javascript',
                              dangerouslySetInnerHTML: {
                                __html: 'window.__routeData = ' + JSON.stringify({
                                  path: route.path,
                                  initialProps: initialProps
                                })
                              }
                            }),
                            _react2.default.createElement('script', { async: true, src: '/app.js' })
                          );
                        };

                        html = '<!DOCTYPE html>' + (0, _server.renderToString)(_react2.default.createElement(
                          HtmlTemplate,
                          { staticMeta: staticMeta, Html: Html, Head: Head, Body: Body },
                          _react2.default.createElement('div', { id: 'root', dangerouslySetInnerHTML: { __html: appHtml } })
                        ));


                        html = html.replace(/(href=["'])(\/)/gm, ('$1' + (config.siteRoot ? config.siteRoot + '/' : '/')).replace(/\/{2,}/g, '/'));

                        htmlFilename = _path2.default.join(_paths.DIST, route.path, 'index.html');
                        initialPropsFilename = _path2.default.join(_paths.DIST, route.path, 'routeData.json');
                        writeHTML = _fsExtra2.default.outputFile(htmlFilename, html);
                        writeJSON = _fsExtra2.default.outputFile(initialPropsFilename, JSON.stringify({
                          initialProps: initialProps
                        }));
                        _context.next = 45;
                        return Promise.all([writeHTML, writeJSON]);

                      case 45:
                      case 'end':
                        return _context.stop();
                    }
                  }
                }, _callee, undefined);
              }));

              return function (_x2) {
                return _ref3.apply(this, arguments);
              };
            }())));

          case 4:
          case 'end':
            return _context2.stop();
        }
      }
    }, _callee2, undefined);
  }));

  return function writeRoutesToStatic(_x) {
    return _ref.apply(this, arguments);
  };
}();

exports.writeRoutesToStatic = writeRoutesToStatic;
;

var _temp = function () {
  if (typeof __REACT_HOT_LOADER__ === 'undefined') {
    return;
  }

  __REACT_HOT_LOADER__.register(defaultEntry, 'defaultEntry', 'src/static.js');

  __REACT_HOT_LOADER__.register(getConfig, 'getConfig', 'src/static.js');

  __REACT_HOT_LOADER__.register(writeRoutesToStatic, 'writeRoutesToStatic', 'src/static.js');

  __REACT_HOT_LOADER__.register(buildXMLandRSS, 'buildXMLandRSS', 'src/static.js');
}();

;
